### DISCUSSION TITLE
Can the on_config be used to patch in raw config files?

### DISCUSSION THREAD

**Discussion by couling** [mkdocs/mkdocs#3261] [Q&A]
Is there a way to for a plugin to patch config from mkdocs syntax plugin files?  All attempts I've made so far have resulted in errors.

The use case here is sharing look&feel configuration between 100+ repos.

It seems that various steps in the flow are unexpectedly destructive.  For example running `validate` twice will always fail. 

Eg:

```python
    def on_config(self, config, **kwargs):
        for path in reversed(self.config.shared_path):
            with open(path, "r") as file:
                config.load_file(file)
        with open(config.config_file_path, "r") as file:
            config.load_file(file)
        # config.valudate()  ... fails
        # TODO ? config['plugins'].run_event('startup', command='serve', dirty=dirty)
```

If I fully reload a new config I get an error of the form:

```
  File "/Users/philip/Documents/Development/ControlBox/.venv/lib/python3.11/site-packages/mkdocs/commands/serve.py", line 128, in serve
    config['plugins'].run_event('shutdown')
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'list' object has no attribute 'run_event'
```

**Comment by oprypin** [comment]
Good question. At the moment I don't know any way to achieve exactly what the title asks.

Maybe there should be a new event `on_pre_config(config_dict: dict[str, Any])`.

Other than that, maybe one way to achieve what you're asking is https://www.mkdocs.org/user-guide/configuration/#configuration-inheritance

Also generally feels like maybe this shouldn't require writing a plugin, but hmm

**Reply by couling** [reply]
`INHERIT` would be close, though it leaves putting the parent file in place as an exercise for the user.  Maintaining synchronisation of that across a lot of repos seems hard.

That's where the idea for a plugin came in. We could put the file an internal store (S3, intranet, git repo, ...) and have a plugin fetch it on build.

I like the idea of an `on_pre_config` hook.  Although it would create a chicken-and-egg challenge of activating plugins enabled by the patched in config.  It would need a two phase configuration load:
 1. minimally discover plugins and checks for any `on_pre_config` hooks.
 2. re-loaded with `on_pre_config` given chance to patch in changes.

**Reply by oprypin** [reply]
Oh of course, the chicken-and-egg problem. How did that slip my mind

**Reply by couling** [reply]
I guess one way around it might be to have a different form of plugin for this explicit task.  If `INHERIT` were modified to accept a URL as well as a path, then the scheme in the url could be used to select which plugin to download from.

Eg:
 - URL https://example.com/ might have one plugin based on the requests library
 - URL s3://example-bucket/ might have another plugin to based on AWS's boto3 library

This wouldn't necessarily need any configuration, the necessary information would be right there in the URL.
